File: smash/meshing/_meshing.f90
recursive subroutine mask_upstream_cells(data, col, row, colsize, &
& rowsize, dkind, mask)
    
    integer, intent(in) :: colsize, rowsize, col, row
    integer, dimension(rowsize, colsize), intent(in) :: data
    integer, dimension(8), intent(in) :: dkind
    logical, dimension(rowsize, colsize), intent(inout) :: mask
    
    integer, dimension(8) :: dx = [0, -1, -1, -1, 0, 1, 1, 1]
    integer, dimension(8) :: dy = [1, 1, 0, -1, -1, -1, 0, 1]
    integer :: i, col_s, row_s
    
    mask(row, col) = .true.
    
    do i=1, 8
    
        col_s = col + dx(i)
        row_s = row + dy(i)
        
        if (col_s .gt. 0 .and. col_s .le. colsize .and. &
        &   row_s .gt. 0 .and. row_s .le. rowsize) then
        
            if (data(row_s, col_s) .eq. dkind(i)) then
                
                call mask_upstream_cells(data, col_s, row_s, &
                & colsize, rowsize, dkind, mask)
            
            end if
            
        end if
    
    end do
    
end subroutine mask_upstream_cells

subroutine catchment_dln(data, col, row, xres, yres, area, dkind, &
& max_depth, logical_dln, col_otl, row_otl)
    
    integer, dimension(:,:), intent(in) :: data
    integer, intent(inout) :: col, row
    integer, intent(in) :: max_depth
    real, intent(in) :: xres, yres, area
    integer, dimension(8), intent(in) :: dkind
    logical, dimension(size(data, 1), size(data, 2)), intent(out) :: &
    & logical_dln
    integer, intent(out) :: col_otl, row_otl
    
    logical, dimension(size(data, 1), size(data, 2)) :: &
    & logical_dln_loop
    integer :: i, j, col_loop, row_loop, colsize, rowsize
    real :: min_tol, tol
    
    ! Transform from Python to FORTRAN index
    col = col + 1
    row = row + 1
    
    colsize = size(data, 2)
    rowsize = size(data, 1)
    
    min_tol = 1.
    
    do i=-max_depth, max_depth
    
        do j=-max_depth, max_depth
        
            col_loop = col + i
            row_loop = row + j
            logical_dln_loop = .false.
            
            if (col_loop .gt. 0 .and. col_loop .le. colsize .and. &
            &   row_loop .gt. 0 .and. row_loop .le. rowsize) then
        
                call mask_upstream_cells(data, col_loop, row_loop, colsize, &
                & rowsize, dkind, logical_dln_loop)
                
                tol = abs(area - count(logical_dln_loop) &
                & * xres * yres) / area
                
                if (tol .lt. min_tol) then
                
                    min_tol = tol
                    ! Transform from FORTRAN to Python index
                    col_otl = col_loop - 1
                    row_otl = row_loop - 1
                    logical_dln = logical_dln_loop
                    
                end if
            
            end if
        
        end do
    
    end do

end subroutine catchment_dln

recursive subroutine downstream_cell_drained_area(data, col, row, colsize, &
& rowsize, da)
    
    integer, intent(in) :: col, row, colsize, rowsize
    integer, dimension(rowsize, colsize), intent(in) :: data
    integer, dimension(rowsize, colsize), intent(inout) :: da
    
    integer, dimension(8) :: dx = [0, 1, 1, 1, 0, -1, -1, -1]
    integer, dimension(8) :: dy = [-1, -1, 0, 1, 1, 1, 0, -1]
    
    integer :: fd
    
    fd = data(row, col)
    
    ! Check no data
    if (fd .gt. 0) then
    
        ! Check bounds
        if (col + dx(fd) .gt. 0 .and. col + dx(fd) .le. colsize .and. &
        &   row + dy(fd) .gt. 0 .and. row + dy(fd) .le. rowsize) then
            
            ! Check pit
            if (abs(data(row, col) - data(row + dy(fd), col + dx(fd))) &
            & .ne. 4) then
            
                da(row + dy(fd), col + dx(fd)) = &
                & da(row + dy(fd), col + dx(fd)) + 1
                
                call downstream_cell_drained_area(data, col + dx(fd), &
                & row + dy(fd), colsize, rowsize, da)
                
            end if
        
        end if
    
    end if
        
end subroutine downstream_cell_drained_area

subroutine drained_area(data, da)

    integer, dimension(:,:), intent(in) :: data
    integer, dimension(size(data, 1), size(data, 2)), intent(out) :: da
    
    integer :: col, row, colsize, rowsize
    
    da = 1
    rowsize = size(data, 1)
    colsize = size(data, 2)
    
    do row=1, rowsize
    
        do col=1, colsize
            
            call downstream_cell_drained_area(data, col, row, colsize, &
            & rowsize, da)
        
        end do

    end do

end subroutine drained_area

File: smash/meshing/meshing.py
from __future__ import annotations

import errno
import os
import numpy as np
import rasterio

from . import _meshing

__all__ = ["generate_meshing"]


def _xy_to_colrow(x, y, xll, yll, xres, yres):

    col = int((x - xll) / xres)
    row = int((yll - y) / yres)

    return col, row

def _colrow_to_xy(col, row, xll, yll, xres, yres):

    x = int(col * xres + xll)
    y = int(yll - row * yres)

    return x, y

def _trim_zeros_2D(array, shift_value=False):

    for ax in [0, 1]:

        mask = ~(array == 0).all(axis=ax)

        inv_mask = mask[::-1]

        start_ind = np.argmax(mask)

        end_ind = len(inv_mask) - np.argmax(inv_mask)

        if ax == 0:

            scol, ecol = start_ind, end_ind
            array = array[:, start_ind:end_ind]

        else:

            srow, erow = start_ind, end_ind
            array = array[start_ind:end_ind, :]

    if shift_value:
        return array, scol, ecol, srow, erow

    else:
        return array

def _array_to_ascii(array, path, xll, yll, cellsize, no_data_val):

    array = np.copy(array)
    array[np.isnan(array)] = no_data_val
    header = (
        f"NCOLS {array.shape[1]} \nNROWS {array.shape[0]}"
        f"\nXLLCENTER {xll} \nYLLCENTER {yll} \nCELLSIZE {cellsize} \nNODATA_value {no_data_val}\n"
    )

    with open(path, "w") as f:

        f.write(header)
        np.savetxt(f, array, "%5.2f")

def _standardize_generate_meshing(x, y, area, name):
    
    x_array = np.array(x, dtype=np.float32, ndmin=1)
    y_array = np.array(y, dtype=np.float32, ndmin=1)
    area_array = np.array(area, dtype=np.float32, ndmin=1)

    # TODO Add check for size
    
    name_array = np.zeros(shape=(20, len(x_array)), dtype="uint8")
    
    if name is None:
        
        for i in range(len(x_array)):
            
            name_ord = [ord(l) for l in ["_", "c", str(i)]]
            
            name_array[0:3, i] = name_ord
            name_array[3:, i] = 32
            
    else:
        
        for i in range(len(x_array)):
            
            name_array[0:len(name[i]), i] = [ord(l) for l in name[i]]
            name_array[len(name[i]):, i] = 32
            
    return x_array, y_array, area_array, name_array

# TODO update for multi-gauge
def generate_meshing(
    path: str,
    x: (float, list[float]),
    y: (float, list[float]),
    area: (float, list[float]),
    dkind: list[int] = [1, 2, 3, 4, 5, 6, 7, 8],
    max_depth: int = 1,
    code: (None, str) = None,
):

    if os.path.isfile(path):
        ds_flow = rasterio.open(path)

    else:
        raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), path)
        
    (x, y, area, code) = _standardize_generate_meshing(x, y, area, code)
    
    flow = ds_flow.read(1)

    ncol = ds_flow.width
    nrow = ds_flow.height

    transform = ds_flow.transform

    xll = transform[2]
    yll = transform[5]
    xres = transform[0]
    yres = -transform[4]
    
    col_otl = np.zeros(shape=x.shape, dtype=np.int32)
    row_otl = np.zeros(shape=x.shape, dtype=np.int32)
    area_otl = np.zeros(shape=x.shape, dtype=np.float32)
    global_mask_dln = np.zeros(shape=flow.shape, dtype=np.int32)

    for ind in range(len(x)):
        
        col, row = _xy_to_colrow(x[ind], y[ind], xll, yll, xres, yres)
        
        mask_dln, col_otl[ind], row_otl[ind] = _meshing.catchment_dln(flow, col, row, xres, yres, area[ind], dkind, max_depth)
        
        area_otl[ind] = np.count_nonzero(mask_dln == 1)
        
        global_mask_dln = np.where(mask_dln == 1, 1, global_mask_dln)
        
    flow = np.ma.masked_array(flow, mask=(1 - global_mask_dln))
    
    flow, scol, ecol, srow, erow = _trim_zeros_2D(flow, shift_value=True)
    global_mask_dln = _trim_zeros_2D(global_mask_dln)
    
    xll_shifted = xll + scol * xres
    yll_shifted = yll - erow * yres
    
    col_otl = col_otl - scol
    row_otl = row_otl - srow
    
    drained_area = _meshing.drained_area(flow)
    
    drained_area = np.ma.masked_array(drained_area, mask=(1 - global_mask_dln))
    
    global_active_cell = global_mask_dln.astype(np.int32)

    gauge_pos = np.vstack((row_otl + 1, col_otl + 1))
    
    mesh = {
        "ng": len(x),
        "nrow": flow.shape[0],
        "ncol": flow.shape[1],
        "xll": xll_shifted,
        "yll": yll_shifted,
        "flow": flow,
        "drained_area": drained_area,
        "global_active_cell": global_active_cell,
        "local_active_cell": global_active_cell.copy(),
        "gauge_pos": gauge_pos,
        "code": code,
        "area": area_otl,
    }

    return mesh
